### PR TITLE
Make spread tests for spread project run on google backend

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This change how the project is prepared and introduce some tests fixes
and a fix in the code as well to make the residue test work again.

### CODE REVIEW

**Comment by sergiocazzolato** [comment]
@mvo5 @niemeyer Please take a look to the change.
To run the tests just do: spread google:ubuntu-18.04-64:
Thanks

File: tests/qemu/task.yaml [hunk 1]
```diff
@@ -6,8 +6,11 @@ prepare: |
         exit 1
     fi
 
+    apt install -y autopkgtest genisoimage
+    autopkgtest-buildvm-ubuntu-cloud -a amd64 -r xenial
+
     mkdir -p ~/.spread/qemu
-    ln -sf /dev/sda ~/.spread/qemu/ubuntu-16.04.img
+    mv autopkgtest-xenial-amd64.img ~/.spread/qemu/ubuntu-16.04.img
```
Comment (niemeyer) [inline tests/qemu/task.yaml]: What we used to have here would use a cached image. This is building the image every single time we run this test. If we have to debug this, we'll be paying the price of every single time too.

File: spread/runner.go [hunk 2]
```diff
@@ -631,7 +631,7 @@ func (r *Runner) worker(backend *Backend, system *System, order []int) {
 				debug = ""
 				repeat = -1
 			}
-			if !abend && !r.options.Restore && repeat == 0 {
+			if !abend && !r.options.Restore && repeat <= 0 {
```
Comment (niemeyer) [inline spread/runner.go]: It's a very good practice to review one's own suggested change before others do.

File: spread.yaml [hunk 3]
```diff
@@ -61,19 +68,4 @@ prepare: |
...
-    LXD_IPV4_DHCP_MAX="250"
-    LXD_IPV4_NAT="true"
-    END
-        systemctl stop lxd lxd-bridge
-        systemctl start lxd lxd-bridge
-    fi
-
-
-# vim:ts=4:sw=4:et
+# vim:ts=4:sw=4:et
```
Comment (niemeyer) [inline spread.yaml]: Missing EOL.

File: spread.yaml [hunk 4]
```diff
@@ -32,19 +42,16 @@ suites:
...
-    # Cache image with: mkdir tests/cache/lxd-ubuntu-16.04 && lxc export ubuntu:16.04 tests/cache/lxd-ubuntu-16.04
-    if ! lxc image list | grep '16\.04'; then
-        lxc image import tests/cache/lxd-ubuntu-16.04/meta-* tests/cache/lxd-ubuntu-16.04/[a-f0-9]* os=ubuntu version=16.04 architecture=amd64 remote=ubuntu
-    fi
-
+    mkdir -p tests/cache
+    
+    # Start lxd instance
+    lxd init --auto
+    lxc launch ubuntu:16.04
```
Comment (niemeyer) [inline spread.yaml]: The original logic was building a cache and allowing one to use an existing cache. The new one has blown that away. It would be good to preserve the ability to use a cached image, and test it to make sure it still works.  Otherwise every time we run tests with qemu, we'll be paying the price of redownloading an image.

File: spread.yaml [hunk 5]
```diff
@@ -8,15 +8,25 @@
...
+    google:
+        key: "$(HOST: echo $SPREAD_GOOGLE_KEY)"
+        location: computeengine/us-east1-b
+        halt-timeout: 2h
+        systems:
+            - ubuntu-18.04-64:
+                image: ubuntu-1804-64-virt-enabled
+                workers: 1
+
     qemu:
```
Comment (niemeyer) [inline spread.yaml]: Is this backend still working?

File: tests/envs/task.yaml [hunk 6]
```diff
@@ -13,3 +13,6 @@ execute: |
 
 debug: |
     cat task.out || true
+
+restore: |
+    rm -f .spread-reuse.yaml /run/spread-reuse.yaml
```
Comment (niemeyer) [inline tests/envs/task.yaml]: That doesn't sound right.  What's the point of using --reuse if it's blowing the data that allows reuse away?   Why is --reuse there in the first place?

File: tests/qemu/task.yaml (hunk 1 above)
Comment (sergiocazzolato) [reply tests/qemu/task.yaml]: Ok, I was thinking on run this just on google, but we really need also to be able to run that on qemu, I'll add the logic to keep both backends working.
Thanks for reviewing this

File: spread/runner.go (hunk 2 above)
Comment (sergiocazzolato) [reply spread/runner.go]: This change is to make residue work when the test fails, currently it was just leaving the residue when the test passed. I noticed that when I ran the residue test and it was failing.

File: spread.yaml (hunk 4 above)
Comment (sergiocazzolato) [reply spread.yaml]: Sure, I'll restore the cache.
Thanks

File: tests/envs/task.yaml (hunk 6 above)
Comment (sergiocazzolato) [reply tests/envs/task.yaml]: Agree, I need just to delete .spread-reuse.yaml to be able to run with -repeat option.

File: spread.yaml (hunk 5 above)
Comment (sergiocazzolato) [reply spread.yaml]: It is not working for me on bionic with qemu 2.11.1, bu works using qemu on xenial.

File: spread/runner.go (hunk 2 above)
Comment (niemeyer) [reply spread/runner.go]: This seems unrelated to this PR purpose.

File: spread.yaml (hunk 4 above)
Comment (niemeyer) [reply spread.yaml]: Does it still work?

File: spread.yaml (hunk 5 above)
Comment (niemeyer) [reply spread.yaml]: Missing response here.

Comment (sergiocazzolato) [reply spread.yaml]: sorry, qemu still works

File: spread/runner.go (hunk 2 above)
Comment (sergiocazzolato) [reply spread/runner.go]: @niemeyer, Without this change the residue test is going to fail on google and qemu as well. I dont see any problem to revert that and add it in another PR. What do you prefer?

Comment (niemeyer) [reply spread/runner.go]: Ack, thanks for the explanation online.

File: spread.yaml (hunk 4 above)
Comment (sergiocazzolato) [reply spread.yaml]: @niemeyer  Yes, it is working

**Comment by sergiocazzolato** [comment]
@niemeyer, hi, the PR is ready. Thanks

File: spread.yaml [hunk 7]
```diff
@@ -32,18 +45,11 @@ suites:
         summary: Integration tests
 
 prepare: |
-    DEBS="git lxd kvm"
-    if ! dpkg -l $DEBS; then
-        apt-get update
-        apt-get install -y $DEBS
-    fi
+    apt update
+    apt install -y git qemu-kvm lxd tree
```
Comment (niemeyer) [inline spread.yaml]: The original logic here would prevent the time consuming apt-get update/install every time if the debs were already installed. Was there a reason for the modification?

File: spread.yaml [hunk 8]
```diff
@@ -61,19 +67,19 @@ prepare: |
...
-    LXD_IPV4_DHCP_MAX="250"
-    LXD_IPV4_NAT="true"
-    END
-        systemctl stop lxd lxd-bridge
-        systemctl start lxd lxd-bridge
+    # Cache is only available if SEND_CACHE=1 is set, follow the steps:
+    # . Create dest dir: > mkdir tests/cache/lxd-ubuntu-16.04
+    # . Get image id: > lxc image list
+    # . Export image: > lxc image export <IAMGE-ID> tests/cache/lxd-ubuntu-16.04
+    # . Extract matadata: > cd tests/cache/lxd-ubuntu-16.04 && tar -xvf <MATADATA-FILE>
```
Comment (niemeyer) [inline spread.yaml]: s/mata/meta/

Can we please also have "-" as the list item prefix instead of "."

Also, why that tar in the last command?  Seems surprising that `lxc image import` cannot import what `lxc image export` exports.

File: spread.yaml [hunk 9]
```diff
@@ -8,17 +8,30 @@
...
+            - ubuntu-18.04-64:
+                image: ubuntu-1804-64-virt-enabled
+                workers: 1
+
     qemu:
         systems:
             - ubuntu-16.04-64:
                 username: ubuntu
                 password: ubuntu
+            - ubuntu-18.04-64:
```
Comment (niemeyer) [inline spread.yaml]: Are both systems still necessary? If you tested with the latter, we can probably drop the former.

File: spread.yaml [hunk 10]
```diff
@@ -61,19 +67,19 @@ prepare: |
...
+    # . Extract matadata: > cd tests/cache/lxd-ubuntu-16.04 && tar -xvf <MATADATA-FILE>
+    if [ -d tests/cache/lxd-ubuntu-16.04 ]; then
+        if ! lxc image list | grep '16\.04'; then
+            lxc image import tests/cache/lxd-ubuntu-16.04
+        fi
     fi
 
+    # Start lxd instance
+    lxd init --auto
+    lxc launch ubuntu:16.04
```
Comment (niemeyer) [inline spread.yaml]: Why is this required now?

File: spread.yaml (hunk 7 above)
Comment (sergiocazzolato) [reply spread.yaml]: No, I just changed that to get updates. I'll revert that.

File: spread.yaml (hunk 8 above)
Comment (sergiocazzolato) [reply spread.yaml]: Agree, it is super weird but this is a lxd known issue and I found the solution in a forum.

File: spread.yaml (hunk 9 above)
Comment (sergiocazzolato) [reply spread.yaml]: No, I'll leave 18.04, thanks

File: spread.yaml (hunk 10 above)
Comment (sergiocazzolato) [reply spread.yaml]: sorry, is is not required.

**Comment by sergiocazzolato** [comment]
@niemeyer comments addressed.

Thanks for the review

**Comment by mvo5** [comment]
Fwiw, I just ran the tests on qemu and GCE and everything is fine: https://paste.ubuntu.com/p/xR5dWRrmbW/ and
